Simple-Forms - Fix Review-page Edit-mode duplicate H3s from titleUIs

tlei123 opened the pull request:
## Summary

- Fix Review-page Edit-mode duplicate H3s, where a page-uiSchema's using titleUI to render an H3 on the form-page.
  - Because neither titleUI nor uiSchema is a component, I fixed this lower-down, in TitleField which already was receiving state-props.
- Team: Veteran Facing Forms
- Flipper not implemented yet

## Related issue(s)

- _Link to ticket created in va.gov-team-forms repo_
department-of-veterans-affairs/va.gov-team-forms#688

## Testing done

- Local browser-testing and unit-tests were all successful.

## Screenshots

### Before
![image](https://github.com/department-of-veterans-affairs/va.gov-team/assets/587583/a248c233-f9e6-42e9-b621-15c1dfc5a036)

### After
![image](https://github.com/department-of-veterans-affairs/vets-website/assets/587583/567d0227-ecfb-4e27-bf08-f0674978e3bd)


## What areas of the site does it impact?

Any forms under src/applications/simple-forms/ with pages using our v3-web-component titleUI pattern.

## Acceptance criteria

## Requested Feedback

Is this the best way to fix our issue?

tlei123 committed:
Fix local-mock-api-responses for 40-0247

tlei123 committed:
Refactor TitleField to conditionally render title
Render strings as-is, but change element's <h3> to <h4> on Review page

tlei123 committed:
Merge branch 'main' into va-forms/40-0247/688-review-page-dupe-h3

rhasselle-oddball commented on src/platform/forms-system/src/js/fields/TitleField.jsx:
```diff
@@ -1,18 +1,49 @@
 import classNames from 'classnames';
 
-export default function TitleField({ id, title }) {
+export default function TitleField({ id, title, formContext }) {
```
That's cool that we have control to update this part, I didn't realize that. the solution looks smart, but I'm thinking what if the team actually wants to show an h3 as an h3 on the review page? I'm sure there are probably use cases like that.
I'm hesitant of the idea of making this default behavior for every form because of retroactively affecting the existing forms, and also the idea that some teams may actually want the styling to be the same.
Also TitleField would probably affect every 'ui:title' situation, including arrays, or maybe some individual components/widgets using 'ui:title'

What if instead we allowed to modify on a case-by-case basis? but I thought we could already achieve this with ui:reviewField ?  but that doesn't work in this situation?

tlei123 replied:
I can't think of why any consumer would intentionally re-introduce the a11y-bug by duplicating the Review-page accordion H3.  The Review-page accordions are hard-coded H3s, so we're always gonna need the page's TitleField to be H4s on the Review page.

rhasselle-oddball replied:
That's a great point. I didn't fully understand but do now. thanks.  Hmm but I think maybe not all teams use that specific array collapsible for their ending page.

rhasselle-oddball commented on src/platform/forms-system/src/js/fields/TitleField.jsx:
```diff
@@ -1,18 +1,49 @@
+  // Handle title scenarios:
+  // If title is a React element with an <h3>,
+  // it should be made <h4> for the Review page.
+  if (isStringTitle) {
```
- if we go with this logic, could you make this so there is only a conditional for our explicit scenario?  for example if (onReviewPage && title.type ==='h3') ... then do what we want, otherwise use default. 
- and then for the existing isEmptyTitle scenario, do we still want to return null instead of a legend with null?

rhasselle-oddball commented on src/platform/forms-system/src/js/fields/TitleField.jsx:
```diff
@@ -1,18 +1,49 @@
+    reviewTitle = !isEmptyStringTitle ? title : null;
+  } else if (!onReviewPage) {
+    reviewTitle = title;
+  } else if (title.type !== 'h3') {
```
hm the other tricky thing about this is:
✅ this works for 'ui:title': <h3>...</h3>,  
X    but not 'ui:title': <> <h3>...</h3></> which is what our titleUI does. 
i'm not sure the best course of action but just thinking about it.

tlei123 replied:
oh, good catch, Robert!  i'll try refactoring for recursive search and replace.

tlei123 replied:
Refactor done; I'm checking for children recursively now and lowering any header-tag's level by one for the Review page.

tlei123 committed:
Merge branch 'main' into va-forms/40-0247/688-review-page-dupe-h3

tlei123 committed:
Refactor TitleField
Modify title React-fragment manipulation code to handle nested/multiple children

tlei123 committed:
Merge branch 'main' into va-forms/40-0247/688-review-page-dupe-h3

tlei123 commented on src/platform/forms-system/src/js/fields/TitleField.jsx:
```diff
@@ -1,18 +1,79 @@
+      return element.map((child, index) => (
+        <React.Fragment key={index}>{getReviewTitle(child)}</React.Fragment>
+      ));
+      // eslint-disable-next-line no-else-return
```
Yes, this is intentional & required because there's a 3rd return outside this sub-block.

tlei123 committed:
Refactor TitleField
Move off Review-page title-manipulation method to helpers

tlei123 commented:
Closing this PR un-merged.  Team decided to take a different approach to fixing this bug.
